### PR TITLE
fix: addresses issue with state diff recording idx

### LINKED ISSUE #1064: removal index (is 360) should be < len (is 360)
### Component

Forge

### Have you ensured that all of these are up to date?

- [x] Foundry
- [x] Foundryup

### What version of Foundry are you on?

forge Version: 1.0.0-dev Commit SHA: 07840ec4e3f68643bacfc3ce234d3242975bb1b3 Build Timestamp: 2025-06-03T09:20:11.176984000Z (1748942411) Build Profile: release

### What version of Foundryup are you on?

_No response_

### What command(s) is the bug in?

forge test

### Operating System

macOS (Apple Silicon)

### System Specifications

_No response_

### Describe the bug

To replicated checkout to this [repo](https://github.com/AaveChan/aave-proposals-v3/blob/ce6d63d949d7e20ae040c18c181fea998b228b54/zksync/src/20250408_AaveV3ZkSync_OnboardWrsETHToZKsyncV3Instance/AaveV3ZkSync_OnboardWrsETHToZKsyncV3Instance_20250408.t.sol) and run `FOUNDRY_PROFILE=zksync forge test --zksync --match-path=zksync/src/20250408_AaveV3ZkSync_OnboardWrsETHToZKsyncV3Instance/AaveV3ZkSync_OnboardWrsETHToZKsyncV3Instance_20250408.t.sol -vv`.
(edit: fixed the repo link)

The zkSolc compiler version used is `v1.5.13`

We get the following error:

```sh
The application panicked (crashed).
Message:  removal index (is 360) should be < len (is 360)
Location: crates/strategy/zksync/src/cheatcode/runner/mod.rs:938

This is a bug. Consider reporting it at https://github.com/matter-labs/foundry-zksync

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                                ⋮ 9 frames hidden ⋮                               
  10: alloc::vec::Vec<T,A>::remove::assert_failed::h77ce4e8bb30dfdda
      at <unknown source file>:<unknown line>
  11: <foundry_strategy_zksync::cheatcode::runner::ZksyncCheatcodeInspectorStrategyRunner as foundry_cheatcodes::strategy::CheatcodeInspectorStrategyExt>::zksync_remove_duplicate_account_access::h37f01b530541c441
      at <unknown source file>:<unknown line>
  12: <foundry_cheatcodes::inspector::Cheatcodes as revm::inspector::Inspector<&mut dyn foundry_evm_core::backend::DatabaseExt+Error = foundry_fork_db::error::DatabaseError>>::call_end::hcc85acd6ac8fc583
      at <unknown source file>:<unknown line>
  13: <foundry_evm::inspectors::stack::InspectorStackRefMut as revm::inspector::Inspector<&mut dyn foundry_evm_core::backend::DatabaseExt+Error = foundry_fork_db::error::DatabaseError>>::call_end::h26e567cc20fe1c6e
      at <unknown source file>:<unknown line>
  14: <foundry_evm::inspectors::stack::InspectorStack as revm::inspector::Inspector<&mut dyn foundry_evm_core::backend::DatabaseExt+Error = foundry_fork_db::error::DatabaseError>>::call_end::h28da02ff3ae50a89
      at <unknown source file>:<unknown line>
  15: revm::inspector::handler_register::inspector_handle_register::{{closure}}::h3a9da57d64072041
      at <unknown source file>:<unknown line>
  16: foundry_evm_core::utils::create2_handler_register::{{closure}}::h06fad5ff51af9c42
      at <unknown source file>:<unknown line>
  17: revm::evm::Evm<EXT,DB>::transact_preverified_inner::h416f6af699eccd87
      at <unknown source file>:<unknown line>
  18: revm::evm::Evm<EXT,DB>::transact::hf562acbdc74a9341
      at <unknown source file>:<unknown line>
  19: <foundry_evm_core::backend::strategy::EvmBackendStrategyRunner as foundry_evm_core::backend::strategy::BackendStrategyRunner>::inspect::h9d21722b5d00ce19
      at <unknown source file>:<unknown line>
  20: <foundry_evm::executors::strategy::EvmExecutorStrategyRunner as foundry_evm::executors::strategy::ExecutorStrategyRunner>::call::h4ea1489b0ec724e1
      at <unknown source file>:<unknown line>
  21: foundry_evm::executors::Executor::call_with_env::h199f8f97fe34260f
      at <unknown source file>:<unknown line>
  22: foundry_evm::executors::Executor::call::h1df517f477066373
      at <unknown source file>:<unknown line>
  23: forge::runner::FunctionRunner::run::hb2a0bed30b060b44
      at <unknown source file>:<unknown line>
  24: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &F>::call_mut::h424ee5dfa8724100
      at <unknown source file>:<unknown line>
  25: rayon::iter::plumbing::Folder::consume_iter::h701364864b89030a
      at <unknown source file>:<unknown line>
  26: rayon::iter::plumbing::bridge_producer_consumer::helper::hb89a56df30404304
      at <unknown source file>:<unknown line>
  27: rayon_core::join::join_context::{{closure}}::hcd297e10ae4992b0
      at <unknown source file>:<unknown line>
  28: <rayon_core::job::StackJob<L,F,R> as rayon_core::job::Job>::execute::h89af56bf6748c1a8
      at <unknown source file>:<unknown line>
  29: rayon_core::registry::WorkerThread::wait_until_cold::hd1c14738ce310561
      at <unknown source file>:<unknown line>
  30: rayon_core::registry::ThreadBuilder::run::h80ad7e474a0188d7
      at <unknown source file>:<unknown line>
  31: std::sys::backtrace::__rust_begin_short_backtrace::h09b4679774d7bde1
      at <unknown source file>:<unknown line>
  32: core::ops::function::FnOnce::call_once{{vtable.shim}}::hbe28125a9243e9bd
      at <unknown source file>:<unknown line>
  33: std::sys::pal::unix::thread::Thread::new::thread_start::h6d53b1b0c047a3b9
      at <unknown source file>:<unknown line>
  34: __pthread_deallocate<unknown>
      at <unknown source file>:<unknown line>
```

### PR DESCRIPTION
**Pull request by dutterbutter**
# What :computer: 
* Updates `zksync_remove_duplicate_account_access` to use the most recent stack entry and check bounds before removal, logging a warning if the index is out of range

# Why :hand:
* Closes #1064 

# Evidence :camera:
Include screenshots, screen recordings, or `console` output here demonstrating that your changes work as intended

# Documentation :books:
Please ensure the following before submitting your PR:

- [ ] Check if these changes affect any documented features or workflows.
- [ ] Update the [book](https://github.com/matter-labs/foundry-zksync-book) if these changes affect any documented features or workflows.

### CODE REVIEW

File: crates/strategy/zksync/src/cheatcode/runner/mod.rs [hunk 1]
```diff
@@ -78,7 +78,9 @@ impl ZksyncCheatcodeInspectorStrategyRunner {
             // If we have a pending stack, it will be appended to the end of primary stack, else at
             // the beginning, once the record is finalized.
             let stack_insert_index = if recorded_account_diffs_stack.len() > 1 {
-                recorded_account_diffs_stack.first().map_or(0, Vec::len)
+                recorded_account_diffs_stack
+                    .get(recorded_account_diffs_stack.len() - 2)
+                    .map_or(0, Vec::len)
```
[the remark this answers is not in the document]
Comment (dutterbutter) [reply crates/strategy/zksync/src/cheatcode/runner/mod.rs:84]: Updated the comment accordingly.

**Comment by dutterbutter** [comment]
> Do we have a small reproducible example to be used as a test?

Sorta, we have the confirmation that it resolves the posted issue in https://github.com/matter-labs/foundry-zksync/issues/1064#issuecomment-2971328405 and the existing test cases pass. However, given this is an edge case and we are still debugging https://github.com/matter-labs/foundry-zksync/issues/1073 which depends on these changes, I will move to a draft and follow up with a test.

**Comment by dutterbutter** [comment]
@popzxc updated comment and added unit test for specific use case

**Review by popzxc — APPROVED** [review]
